sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkExecuteStatementOperation.scala

```diff
@@ -111,9 +115,15 @@ private[hive] class SparkExecuteStatementOperation(
 
     // Reset iter to header when fetching start from first row
     if (order.equals(FetchOrientation.FETCH_FIRST)) {
-      val (ita, itb) = iterHeader.duplicate
-      iter = ita
-      iterHeader = itb
+      iter = if (useIncrementalCollect) {
+        resultList = None
+        result.toLocalIterator.asScala
+      } else {
+        if (resultList.isEmpty) {
```

Comment:
OK I believe I get it now. I see your approach and it makes sense. 
The only real change here is that you hold on to a reference to the whole data set here rather than `collect()` it into memory. Maybe that's the right thing to do but that's the only thing I'm wondering about. Previously it seems like it would `collect()` each time anyway?

Just wondering if that's actually simpler, to avoid keeping a reference to the whole data set, or whether that defeats a purpose.